pyro/contrib/funsor/handlers/plate_messenger.py

```diff
@@ -154,3 +156,170 @@ def _pyro_param(self, msg):
+        #       log_prob   3 1 1 1 1 |
+        #       x_1 dist   3 1 1 1 1 |
+        #          value 3 1 1 1 1 1 |
+        #       log_prob 3 3 1 1 1 1 |
+        #       y_1 dist 3 1 1 1 1 1 |
+        #          value             |
+        #       log_prob 3 1 1 1 1 1 |
+        #       x_2 dist 3 1 1 1 1 1 |
+        #          value   3 1 1 1 1 |
+        #       log_prob 3 3 1 1 1 1 |
+        #       y_2 dist   3 1 1 1 1 |
+        #          value             |
+        #       log_prob   3 1 1 1 1 |
+
+    .. warning::  This is only valid if there is only one Markov
+        dimension per branch.
+
+    :param str name: A unique name of a Markov dimension to help inference algorithm
+        eliminate variables in the Markov chain.
+    :param int size: Length (size) of the Markov chain.
+    :param int dim: An optional dimension to use for this independence index.
+        If specified, ``dim`` should be negative, i.e. should index from the
+        right. If not specified, ``dim`` is set to the rightmost dim that is
+        left of all enclosing ``plate`` contexts.
+    :param int history: Memory (order) of the Markov chain. Also the number
+        of previous contexts visible from the current context. Defaults to 1.
+        If zero, this is similar to :class:`pyro.plate`.
+    :return: Returns both `int` and 1-dimensional :class:`torch.Tensor` indices:
+        (`0, ..., history-1, torch.arange(size-history), ..., torch.arange(history, size)`).
+    """
+    def __init__(self, name=None, size=None, dim=None, history=1):
+        self.name = name
+        self.size = size
+        self.dim = dim
+        self.history = history
+        super().__init__()
+
+    @staticmethod
+    @effectful(type="markov_chain")
+    def _markov_chain(name=None, markov_vars=set(), suffixes=list()):
+        """
+        Constructs names of markov variables in the `chain`
+        from markov_vars prefixes and suffixes.
+
+        :param str name: The name of the markov dimension.
+        :param set markov_vars: Markov variable name markov_vars.
+        :param list suffixes: Markov variable name suffixes.
+            (`0, ..., history-1, torch.arange(0, size-history), ..., torch.arange(history, size)`)
+        :return: step information
+        :rtype: frozenset
+        """
+        chain = frozenset({tuple("{}{}".format(var, suffix) for suffix in suffixes)
+                          for var in markov_vars})
+        return chain
+
+    def __iter__(self):
+        self._auxiliary_to_markov = {}
+        self._markov_vars = set()
+        with self:
+            with IndepMessenger(name=self.name, size=self.size-self.history, dim=self.dim) as time:
+                self._suffixes = \
+                    list(range(self.history)) \
+                    + list(time.indices+i for i in range(self.history+1))
+                for self._suffix in MarkovMessenger(history=self.history)(self._suffixes):
+                    yield self._suffix
```

Comment:
Wow, this is awesome! Great that it turned out to be so simple!

Reply:
There was one issue. Had to take out `range(self.history)` initialization steps out of `IndepMessenger`, otherwise `time` dim gets added to the `cond_indep_stack` and broadcasted. Also removed `MarkovMessenger` for now because there are two for loops.